### PR TITLE
[decompositions] GRU decompositon with and without packed sequence

### PR DESCRIPTION
**Pull request by samdow**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #91466
* #91465
* #91281
* #91124
* #91807
* #91123

### CODE REVIEW

File: torch/_decomp/decompositions.py [hunk 1]
```diff
@@ -2513,6 +2535,84 @@ def lstm_data_impl(
...
+    return (cur_hidden - new_gate) * input_gate + new_gate
+
+
+def gru_helper_data(inp, cur_hidden, ih_weight, ih_bias, hh_weight, hh_bias):
+    chunked_igates = F.linear(inp, ih_weight, ih_bias).chunk(3, 1)
+    chunked_hgates = F.linear(cur_hidden, hh_weight, hh_bias).chunk(3, 1)
+    reset_gate = (chunked_hgates[0] + chunked_igates[0]).sigmoid()
+    input_gate = (chunked_hgates[1] + chunked_igates[1]).sigmoid()
+    new_gate = (chunked_igates[2] + (chunked_hgates[2] * reset_gate)).tanh()
+    return (cur_hidden - new_gate) * input_gate + new_gate
```
Comment (zou3519) [inline torch/_decomp/decompositions.py:2553]: nit: correct me if I'm wrong, but all the helper functions basically look like the cells (e.g. rnn_cell, gru_cell). We could name them that if that sounds reasonable, but not a big deal

**Review by zou3519 — APPROVED** [review]
Cool!

**Comment by malfet** [comment]
@lezcano teach me the trick how you've managed to push changes to other people's stack (and now it looks like all the changes are authored by you rather than by @samdow which is a bit weird)

**Comment by lezcano** [comment]
I simply picked the branch `gh/samdow/58/orig`, which contains all the ghstack commits and created a new branch. Then, I started editing this new branch and run `ghstack`  on it. As `ghstack` just checks the metadata in the commit messages to figure out where to push things, ghstack correctly updates this stack of PRs.

And, ugh, the authorship part is nasty. I figured it'd be shown as authored by both. This may be a bug in ghstack? @ezyang

**Comment by lezcano** [comment]
Actually, I don't think it's a bug but a corollary of how I did it.. But I don't know how to do this trick but also keep the authorship from both... :/

**Pull request closed without merging**